Sharding Guide (#9562)
Rapptz/discord.py: docs/guide/topics/sharding.rst

.. currentmodule:: discord

.. _guide_sharding:

Sharding
==========

When bots start to get considerably large, the amount of events they have to deal with can start to become problematic.

At high user and guild counts the incoming messages, typing events, and status updates can start to climb to being as frequent as hundreds per second.

To help deal with this large amount of traffic, Discord supports **sharding**, a feature where your bot can split its guilds amongst separate connections, reducing the amount of data each individual connection has to handle.

This not only helps Discord by reducing how much data they need to direct to one place, but also helps us, as we can split our bot's overall work across different connections, environments, or even entirely different machines. 

Discord recommends using sharding once you get over 1,000 guilds, and once you reach beyond 2,500 guilds, it becomes a requirement.

> Now, let's discuss what options we have for setting up sharding within discord.py.

Reviewer: The sentences in this introduction section are a bit long. This makes the paragraph harder to comprehend, so it may be helpful if the sentences are broken up.

Specifying shard count and IDs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You may want to specify a total shard count instead of relying on Discord's recommendation if you find you need more or less shards, or if you want to keep it consistent across restarts of your bot.

.. warning::

    When specifying a shard count, note that each shard must have less than 2,500 guilds.

Specifying shard IDs is useful for bots running as multiple processes. For example, if a bot has 16 shards, you may have one process run shards 0-7 and another process run shards 8-15. These values can be specified with, for example, an environment variable, to allow passing different values to each process. If you specify shard IDs, you must also specify a shard count.

.. note::

    This does not have the same effect vice versa.

Examples
~~~~~~~~~~

Let's make a bot using :class:`~ext.commands.AutoShardedBot`:

.. code-block:: python3

    import discord
    from discord.ext import commands

>     bot = commands.AutoShardedBot(command_prefix='!')

Reviewer 2: Missing intents here
```suggestion
    intents = discord.Intents.default()
    intents.message_content = True
    bot = commands.AutoShardedBot(command_prefix='!', intents=intents)
```

Author: thank you!

    @bot.command()
    async def shards(ctx):
        await ctx.send(f"I am running on {bot.shard_count} shards!")
    
    bot.run("token")

.. note::
    
    You can specify the shard_count in AutoShardedBot, but that defeats the whole purpose of :class:`~AutoShardedBot` and you should just use :class:`~Bot` instead.

If you don't wanna use discord's recommended shard count, you can specify your own:

.. code-block:: python3

    import discord
    from discord.ext import commands

>     bot = commands.Bot(command_prefix='!', shard_count=10)

Reviewer 2: Missing intents here also
```suggestion
    intents = discord.Intents.default()
    intents.message_content = True
    bot = commands.Bot(command_prefix='!', shard_count=10, intents=intents)
```

.. currentmodule:: discord

.. _guide_sharding:

Sharding
==========

When bots start to get considerably large, the amount of events they have to deal with can start to become problematic.

> At high user and guild counts the incoming messages, typing events, and status updates can start to climb to being as frequent as hundreds per second.

Author: I will rewrite on my next commit, thanks for pointing it out.

To help deal with this large amount of traffic, Discord supports **sharding**, a feature where your bot can split its guilds amongst separate connections, reducing the amount of data each individual connection has to handle.

> This not only helps Discord by reducing how much data they need to direct to one place, but also helps us, as we can split our bot's overall work across different connections, environments, or even entirely different machines. 

Author: I'll remove on the next commit, thanks for your review ❤️ @fretgfr

Discord recommends using sharding once you get over 1,000 guilds, and once you reach beyond 2,500 guilds, it becomes a requirement.

Now, let's discuss what options we have for setting up sharding within discord.py.

Client vs AutoShardedClient
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Both :class:`~Client` and :class:`~ext.commands.Bot` have auto sharded variants, :class:`~AutoShardedClient` and :class:`~ext.commands.AutoShardedBot` respectively.

> The key difference between the two is that the former can only support one connection (one shard), while the auto sharded variants can handle multiple gateway connections. If you are running multiple shards, it is recommended to use the auto sharded variants and have multiple connections in each process.

Author: will delete at the end, doesn't really make sense being there.

> The key difference between the two is that the former can only support one connection (one shard), while the auto sharded variants can handle multiple gateway connections. If you are running multiple shards, it is recommended to use the auto sharded variants and have multiple connections in each process.

Author: >"The key difference between the two is that the former can only support one connection (one shard), while the auto sharded variants can handle multiple gateway connections. If you are running multiple shards, it is recommended to use the auto sharded variants."

I don't think this makes sense, since afaik commands.Bot suppports shard_count and shard_ids. can someone correct me on this?

> The key difference between the two is that the former can only support one connection (one shard), while the auto sharded variants can handle multiple gateway connections. If you are running multiple shards, it is recommended to use the auto sharded variants and have multiple connections in each process.

Author: @fretgfr corrected me, they do not. closing convo and altering text.

There's 2 ways you can do sharding:

Using auto sharding
~~~~~~~~~~~~~~~~~~~~~

If you want to use a single process and the recommended number of shards by Discord, you can simply use :class:`~AutoShardedClient` or :class:`~ext.commands.AutoShardedBot` instead of :class:`~Client` or :class:`~ext.commands.Bot` respectively.

Specifying shard count and IDs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You may want to specify a total shard count instead of relying on Discord's recommendation if you find you need more or less shards, or if you want to keep it consistent across restarts of your bot.

.. warning::

    When specifying a shard count, note that each shard must have less than 2,500 guilds.

> 

Reviewer 3: It doesn't need to be much, but I do think it's worth mentioning in a conspicuous place.

> 

Author: will add next commit

Specifying shard IDs is useful for bots running as multiple processes. For example, if a bot has 16 shards, you may have one process run shards 0-7 and another process run shards 8-15. These values can be specified with, for example, an environment variable, to allow passing different values to each process. If you specify shard IDs, you must also specify a shard count.

.. note::

>     This does not have the same effect vice versa.

Reviewer 3: I think the "If you specify shard IDs you must also specify a shard count" belongs in the note alongside the other piece of info, not above.

Maybe an explicit mention that shard ids are zero indexed as well.

Examples
~~~~~~~~~~

Let's make a bot using :class:`~ext.commands.AutoShardedBot`:

.. code-block:: python3

    import discord
    from discord.ext import commands

    intents = discord.Intents.default()
    intents.message_content = True
    bot = commands.AutoShardedBot(command_prefix='!', intents=intents)

    @bot.command()
    async def shards(ctx):
        await ctx.send(f"I am running on {bot.shard_count} shards!")
    
    bot.run("token")

.. note::
    

>     You can specify the shard_count in AutoShardedBot, but that defeats the whole purpose of :class:`~AutoShardedBot` and you should just use :class:`~Bot` instead.

Author: I can phrase it better, this isn't that good.

> If you don't wanna use discord's recommended shard count, you can specify your own:

Author: fair, overlooked.

Discord recommends sharding when your application reaches 1,000 guilds. Once your application reach 2,500 guilds, sharding is required.

Let's discuss our options for setting up sharding with discord.py.

Client vs AutoShardedClient
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Both :class:`~Client` and :class:`~ext.commands.Bot` have auto sharded variants, :class:`~AutoShardedClient` and :class:`~ext.commands.AutoShardedBot` respectively.

The key difference between the two is that the former can only support one connection (one shard), while the auto sharded variants can handle multiple gateway connections.

There's 2 ways you can do sharding:

Using auto sharding
~~~~~~~~~~~~~~~~~~~~~

If you want to use a single process and the recommended number of shards by Discord, you can simply use :class:`~AutoShardedClient` or :class:`~ext.commands.AutoShardedBot` instead of :class:`~Client` or :class:`~ext.commands.Bot` respectively.

Specifying shard count and IDs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You may want to specify a total shard count instead of relying on Discord's recommendation if you find you need more or less shards, or if you want to keep it consistent across restarts of your bot.

Reviewer 4: ```suggestion
You may want to specify a total shard count instead of relying on Discord's recommendation to regulate load balancing, if for example, your clusters do not have the computational requirements to handle the recommended number.

.. note::

    The shard count does not restrict the number of potential shards you have as it is only used for routing traffic. In other words, you can have multiple sessions running as shard ``3`` for example, allowing you to orchestrate a "zero-downtime" solution.
```

.. note::

>     When nearing 150,000 guilds, Discord will migrate your bot to [large bot sharding](https://discord.com/developers/docs/topics/gateway#sharding-for-large-bots). More details are available on their documentation page.

Reviewer 5: This isn't how you do hyperlinks in RST, it's like so:

```suggestion
    When nearing 150,000 guilds, Discord will migrate your bot to `large bot sharding <https://discord.com/developers/docs/topics/gateway#sharding-for-large-bots>`_. More details are available on their documentation page.
```

.. warning::

    When specifying a shard count, note that each shard must have less than 2,500 guilds.

> Specifying shard IDs is useful for bots running as multiple processes. For example, if a bot has 16 shards, you may have one process run shards 0-7 and another process run shards 8-15. These values can be specified with, for example, an environment variable, to allow passing different values to each process.

Reviewer 4: ```suggestion
Specifying shard IDs is useful for bots running as multiple processes. For example, if a bot has 16 shards, you may have one process run shards 0-7 and another process run shards 8-15. These values can be specified with, for example, an environment variable, to allow passing different values to each process. This behaviour can be achieved with ``commands.AutoShardedBot`` and its ``shard_ids`` parameter which accepts a list of values, such as ``list(range(8))``
```

> 

Reviewer 4: ```suggestion

To calculate which shard receives which events, we can use the following formula ``shard_id = (guild_id >> 22) % num_shards``. For example, this may be used to calculate which cluster to send a payload to along an IPC which connects shards and a web dashboard.

```

.. note::

    If you specify shard IDs, you must also specify a shard count. This does not do the same effect vice versa.

.. warning::

    Don't forget that indexing in Python starts at 0, and that shard_id 0 is a valid thing.

Examples
~~~~~~~~~~

Let's make a bot using :class:`~ext.commands.AutoShardedBot`:

.. code-block:: python3

    import discord
    from discord.ext import commands

    intents = discord.Intents.default()
    intents.message_content = True
    bot = commands.AutoShardedBot(command_prefix='!', intents=intents)

    @bot.command()
    async def shards(ctx):
        await ctx.send(f"I am running on {bot.shard_count} shards!")
    
    bot.run("token")

If you don't want to use discord's recommended shard count, you can specify your own:

.. code-block:: python3

    import discord
    from discord.ext import commands

    intents = discord.Intents.default()
    intents.message_content = True
    bot = commands.AutoShardedBot(command_prefix='!', shard_count=10, intents=intents)

    @bot.command()
    async def shards(ctx):
        await ctx.send(f"I am running on {bot.shard_count} shards!")
    
    bot.run("token")

.. note::
    

>     You can specify the shard_count in AutoShardedBot, but it makes more sense to use :class:`~ext.commands.Bot` for this.

Reviewer 5: I'm wondering about the need for this line at all?

For instance upon reading about AutoSharded<Bot> why is this line necessary. Why does it make more sense to use Bot than AutoShardedBot? Needs extrapolation IMO.

Reviewer 4: This line isn't correct. `AutoShardedBot` will create the `shard_ids` through `range(shard_count)` whereas `Bot` will ignore the `shard_count` if a `shard_id` is not provided.
In this case, `AutoShardedBot` is the only valid option.
```suggestion
```
